nifi-extension-bundles/nifi-github-bundle/nifi-github-extensions/src/main/java/org/apache/nifi/github/GitHubFlowRegistryClient.java

```diff
@@ -0,0 +1,633 @@
+        snapshotMetadata.setBucketIdentifier(bucketId);
+    }
+
+    // Ensures the snapshot has non-null flow and metadata fields, which would only be null if taking a flow from "Download Flow Definition" and adding directly to GitHub
+    private void populateFlowAndSnapshotMetadata(final RegisteredFlowSnapshot flowSnapshot, final FlowLocation flowLocation) {
+        if (flowSnapshot.getFlow() == null) {
+            final RegisteredFlow registeredFlow = new RegisteredFlow();
+            registeredFlow.setName(flowLocation.getFlowId());
+            registeredFlow.setIdentifier(flowLocation.getFlowId());
+            flowSnapshot.setFlow(registeredFlow);
+        }
+        if (flowSnapshot.getSnapshotMetadata() == null) {
+            final RegisteredFlowSnapshotMetadata snapshotMetadata = new RegisteredFlowSnapshotMetadata();
+            snapshotMetadata.setFlowIdentifier(flowLocation.getFlowId());
+            flowSnapshot.setSnapshotMetadata(snapshotMetadata);
+        }
+    }
+
+    private synchronized GitHubRepositoryClient getRepositoryClient(final FlowRegistryClientConfigurationContext context) throws IOException, FlowRegistryException {
+        if (!initialized.get()) {
+            getLogger().info("Initializing GitHub repository client");
+            repositoryClient = GitHubRepositoryClient.builder()
+                    .apiUrl(context.getProperty(GITHUB_API_URL).getValue())
+                    .authenticationType(GitHubAuthenticationType.valueOf(context.getProperty(AUTHENTICATION_TYPE).getValue()))
+                    .personalAccessToken(context.getProperty(PERSONAL_ACCESS_TOKEN).getValue())
+                    .appInstallationToken(context.getProperty(APP_INSTALLATION_TOKEN).getValue())
+                    .repoOwner(context.getProperty(REPOSITORY_OWNER).getValue())
+                    .repoName(context.getProperty(REPOSITORY_NAME).getValue())
+                    .repoPath(context.getProperty(REPOSITORY_PATH).getValue())
+                    .build();
+            initialized.set(true);
+
+            // If the client is configured to authenticate, then ensure the directory for the default bucket is present and if not create it,
+            // otherwise the client can only be used to import flows from the repo and won't be able to set up the default bucket
+            if (repositoryClient.getAuthenticationType() != GitHubAuthenticationType.NONE) {
+                final String branch = context.getProperty(REPOSITORY_BRANCH).getValue();
+                final Optional<String> defaultBucketKeepFileSha = repositoryClient.getContentSha(DEFAULT_BUCKET_KEEP_FILE_PATH, branch);
+                if (defaultBucketKeepFileSha.isPresent()) {
+                    getLogger().info("Found default bucket with SHA {}", defaultBucketKeepFileSha);
+                } else {
+                    getLogger().info("Creating default bucket");
+                    repositoryClient.createContent(
```

Comment:
I am not sure that I would make this a default behaviour. This looks to be a hidden writing operation into an external source might be unexpected. It can even happen that the given user has no right for the operation and the client should only read.

I do not doubt the usefullness of this in some cases but I think it should be guarded by a property, something like "create default branch if not found" where the default value for the property is false.

(I see that it is handled using a log for some cases but I think this might cause more consfusion than expected)

Reply:
I'm not sure if I fully agree. I think this is the easiest way for people to get things working. Otherwise they are going to create the client and go to try and use it from NiFi and see no buckets in the list and be confused about what they should do.

The call to create the bucket is only done if the client is configured with authentication, which in most cases would mean they have access to the repo. If you were connecting to a public repo for importing only, then you wouldn't provide any authentication and this call wouldn't happen. I suppose there is an in between case where its a private repo with read-only access, but not sure. Let me think more about this.

Reply:
I changed the behavior to only create the default bucket when the client has WRITE permissions AND when there are no other existing buckets.